### PR TITLE
BUG: Switched shapes in ValueError msg in DataFrame construct (#20742)

### LINKED ISSUE #20742: DataFrame creation incorrect error message
The problem was already mentioned as part of other issues, but still persists in 0.22

https://github.com/pandas-dev/pandas/issues/8020
https://github.com/blaze/blaze/issues/466

Reported both expected shape and input data shape are both transposed which causes a lot of confusion. In my opinion, the reference value should be ` DataFrame.shape`.

`my_arr = np.array([1, 2, 3])
print("my_arr.shape: {}".format(my_arr.shape))
df = pd.DataFrame(index=[0], columns=range(0, 4), data=my_arr)`

`my_arr.shape: (3,)
Traceback (most recent call last):
...
ValueError: Shape of passed values is (1, 3), indices imply (4, 1)
`

Below are shapes which are expected to be reported:
`my_arr = np.array([[0, 1, 2, 3]])
print("my_arr.shape: {}".format(my_arr.shape))
df = pd.DataFrame(index=[0], columns=range(0, 4), data=my_arr)
print(**df.shape**)`

`my_arr.shape: (1, 4)
**(1, 4)**`

### PR DESCRIPTION
**Pull request by meiermark**
- [x] closes #20742
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.rst [hunk 1]
```diff
@@ -1852,6 +1852,7 @@ Other
 ^^^^^
 
 - Bug where C variables were declared with external linkage causing import errors if certain other C libraries were imported before Pandas. (:issue:`24113`)
+- Switched shape in ``ValueError`` message when constructiong a :class:`DataFrame` with parameters ``columns`` and ``index`` not matching the shape of the input data. (:issue:`20742`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.24.0.rst]: Let's be explicit as to how the dimensions are reported (i.e. their order) after the switch.

File: pandas/core/internals/managers.py [hunk 2]
```diff
@@ -1673,8 +1673,10 @@ def create_block_manager_from_arrays(arrays, names, axes):
 
 def construction_error(tot_items, block_shape, axes, e=None):
     """ raise a helpful message about our construction """
-    passed = tuple(map(int, [tot_items] + list(block_shape)))
-    implied = tuple(map(int, [len(ax) for ax in axes]))
+    passed = tuple(map(int, list(block_shape) + [tot_items]))
```
Comment (jreback) [inline pandas/core/internals/managers.py]: just do a reverse if ndim == 2 here

File: pandas/tests/frame/test_constructors.py [hunk 3]
```diff
@@ -386,25 +386,37 @@ def test_constructor_error_msgs(self):
...
 
         # wrong size ndarray, GH 3105
-        msg = r"Shape of passed values is \(3, 4\), indices imply \(3, 3\)"
+        msg = r"Shape of passed values is \(4, 3\), indices imply \(3, 3\)"
         with pytest.raises(ValueError, match=msg):
             DataFrame(np.arange(12).reshape((4, 3)),
                       columns=['foo', 'bar', 'baz'],
                       index=pd.date_range('2000-01-01', periods=3))
 
+        # see issue #20742
```
Comment (jreback) [inline pandas/tests/frame/test_constructors.py]: you don't need the issue number

File: pandas/tests/test_panel.py [hunk 4]
```diff
@@ -1143,23 +1143,11 @@ def test_from_dict_mixed_orient(self):
...
 
     def test_constructor_error_msgs(self):
-        msg = (r"Shape of passed values is \(3, 4, 5\), "
-               r"indices imply \(4, 5, 5\)")
+        msg = (r"Shape of passed values is \(4, 5, 3\), "
+               r"indices imply \(5, 6, 4\)")
         with pytest.raises(ValueError, match=msg):
             Panel(np.random.randn(3, 4, 5),
-                  lrange(4), lrange(5), lrange(5))
-
```
Comment (jreback) [inline pandas/tests/test_panel.py]: leave these tests here

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.rst [hunk 5]
```diff
@@ -1852,6 +1852,9 @@ Other
 ^^^^^
 
 - Bug where C variables were declared with external linkage causing import errors if certain other C libraries were imported before Pandas. (:issue:`24113`)
+- Changed ``ValueError`` message when constructiong a :class:`DataFrame` with parameters ``columns`` and ``index`` not matching the shape of the input data. New message: "Shape of passed values is (``index``, ``columns``), indices imply (``index``, ``columns``)" (:issue:`20742`)
+- Changed ``ValueError`` message when constructiong a :class:`Panel` with parameters ``items``, ``major index`` and ``minor index`` not matching the shape of the input data. New message: "Shape of passed values is (``major``, ``minor``, ``items``), indices imply (``major``, ``minor``, ``items``)" (related to :issue:`20742`)
```
[the remark this answers is not in the document]
Comment (meiermark) [reply doc/source/whatsnew/v0.24.0.rst]: I improved the message. It's now shorter. However, I am not sure if it is short enough or which parts you could still remove.

File: pandas/core/internals/managers.py (hunk 2 above)
Comment (meiermark) [reply pandas/core/internals/managers.py]: Probably not a good idea for panels with 3 dimensions (items, major, minor)
with reversed for ndim == 2: (items, major, minor)
current: (major, minor, items)

The current version is the intended behavior, right?

File: pandas/tests/test_panel.py (hunk 4 above)
Comment (meiermark) [reply pandas/tests/test_panel.py]: okay. Is there additional value in those tests?

File: pandas/tests/frame/test_constructors.py (hunk 3 above)
Comment (meiermark) [reply pandas/tests/frame/test_constructors.py]: okay.

File: pandas/core/internals/managers.py (hunk 2 above)
Comment (jreback) [reply pandas/core/internals/managers.py]: I don't care about Panels, they are being removed shortly.

File: pandas/tests/test_panel.py [hunk 6]
```diff
@@ -1143,20 +1143,20 @@ def test_from_dict_mixed_orient(self):
         assert panel['A'].values.dtype == np.float64
 
     def test_constructor_error_msgs(self):
-        msg = (r"Shape of passed values is \(3, 4, 5\), "
-               r"indices imply \(4, 5, 5\)")
+        msg = (r"Shape of passed values is \(4, 5, 3\), "
```
Comment (jreback) [inline pandas/tests/test_panel.py]: leave panel entirely.

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by TomAugspurger — COMMENTED** [review]
Need to fix a merge conflict.

File: pandas/core/internals/managers.py [hunk 7]
```diff
@@ -1674,7 +1674,13 @@ def create_block_manager_from_arrays(arrays, names, axes):
 def construction_error(tot_items, block_shape, axes, e=None):
     """ raise a helpful message about our construction """
     passed = tuple(map(int, [tot_items] + list(block_shape)))
-    implied = tuple(map(int, [len(ax) for ax in axes]))
+    if len(passed) == 2:
```
Comment (jreback) [inline pandas/core/internals/managers.py]: can you add a comment on the reversing. (e.g. this is a user facing error message and this is a DataFrame construction)

you can do len(passed) <= 2

File: doc/source/whatsnew/v0.24.0.rst [hunk 8]
```diff
@@ -1816,6 +1816,7 @@ Reshaping
 - Bug in :func:`DataFrame.unstack` where a ``ValueError`` was raised when unstacking timezone aware values (:issue:`18338`)
 - Bug in :func:`DataFrame.stack` where timezone aware values were converted to timezone naive values (:issue:`19420`)
 - Bug in :func:`merge_asof` where a ``TypeError`` was raised when ``by_col`` were timezone aware values (:issue:`21184`)
+- Reversed shapes in ValueError message when data shape does not match index shape in construction. (:issue:`20742`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: say Incorrect shape in DataFrame construction error message.

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
whatsnew correction. ping on green.

**Comment by TomAugspurger** [comment]
Thanks @meiermark.

**Pull request merged**